### PR TITLE
add match guards

### PR DESCRIPTION
**Pull request by 1Kinoti**
## description

this pr adds [match guards](https://doc.rust-lang.org/reference/expressions/match-expr.html#match-guards) to match patterns
```nushell
match $x {
   _ if $x starts-with 'nu' => {},
   $x => {}
}
```

these work pretty much like rust's match guards, with few limitations:

1. multiple matches using the `|` are not (yet?) supported
 
```nushell
match $num {
    0 | _ if (is-odd $num) => {},
    _ => {}
}
```

2. blocks cannot be used as guards, (yet?)

```nushell
match $num {
    $x if { $x ** $x == inf } => {},
     _ => {}
}
```

## checklist
- [x] syntax
- [x] syntax highlighting[^1]
- [x] semantics
- [x] tests
- [x] clean up

[^1]: defered for another pr

### CODE REVIEW

**Comment by fdncred** [comment]
oh, this looks cool! can't wait to try it out.

**Comment by 1Kinoti** [comment]
looks like syntax highlighting is going to be a [pain](https://github.com/nushell/nushell/blob/667502e8da0ef8bc06c10f62684c76debae4cf50/crates/nu-parser/src/flatten.rs#L266)

i might defer it to another pr that addresses match patterns as a whole. removing it from the check list for now

**Comment by fdncred** [comment]
i like this pr but it's kind of hard to see the changes. i see changes from several prs here. not sure what git did.

**Comment by 1Kinoti** [comment]
sorry about that, you can click the individual commits named 'ep*' to see specific changes

after all this time, i still haven't fully mastered git :(

**Comment by amtoine** [comment]
i think it's a GitHub issue :eyes: 

if you checkout the tip of this branch, i.e. da7176fe1, and run `git diff --stat main`, i only see 7 files
```
 crates/nu-cmd-lang/src/core_commands/match_.rs | 62 +++++++++++++++++---------
 crates/nu-command/tests/commands/match_.rs     | 51 +++++++++++++++++++++
 crates/nu-parser/src/parse_patterns.rs         |  8 ++++
 crates/nu-parser/src/parser.rs                 | 53 +++++++++++++++++++++-
 crates/nu-protocol/src/ast/expression.rs       |  7 +++
 crates/nu-protocol/src/ast/match_pattern.rs    |  7 ++-
 crates/nu-protocol/src/shell_error.rs          | 12 +++++
 7 files changed, 175 insertions(+), 25 deletions(-)
```

**Comment by amtoine** [comment]
@1Kinoti
maybe merging the current `main` branch into this PR branch will solve this? :shrug:

**Comment by 1Kinoti** [comment]
thanks @amtoine, it worked!

File: crates/nu-parser/src/parser.rs [hunk 1]
```diff
@@ -4208,8 +4208,12 @@ pub fn parse_match_block_expression(working_set: &mut StateWorkingSet, span: Spa
             break;
         }
 
+        let connector = working_set
+            .get_span_contents(output[position].span)
+            .to_vec();
+        let mut connector = connector.as_slice();
```
Comment (jntrnr) [inline crates/nu-parser/src/parser.rs:4214]: I'm not seeing why you needed to turn `connector` into a `vec` and then a slice here. Can you keep it as just a slice?

If it's an ownership thing, you can re-run `get_span_contents` a second time.

Comment (1Kinoti) [reply crates/nu-parser/src/parser.rs]: found a way to remove `to_vec()`

**Comment by fdncred** [comment]
thanks for cleaning up the pr @1Kinoti. it's much easier to see the changes now.

**Comment by jntrnr** [comment]
Looks good

**Pull request merged**